[ENH] Adds extraction of physio signals from DICOMs

pvelasco opened the pull request:
It uses [`bidsphysio`](https://github.com/cbinyu/bidsphysio) to extract physiological signals from CMRR EPI DICOMs.

It includes regression test

pvelasco committed:
[ENH] Adds extraction of physio signals from DICOMs

It uses [`bidsphysio`](https://github.com/cbinyu/bidsphysio) to extract physiological signals from CMRR EPI DICOMs.

It includes regression test

pvelasco committed:
Adds heuristic for physio extraction.

yarikoptic commented on Dockerfile:
```diff
@@ -99,7 +105,8 @@ RUN export PATH="/opt/miniconda-latest/bin:$PATH" \
     &&   pip install --no-cache-dir --editable \
-             '/src/heudiconv[all]'" \
+             '/src/heudiconv[all]' \
+	     '/src/bidsphysio'" \
```
this file is generated, shouldn't be modified directly. there is `utils/gen-docker-image.sh` to be tuned but even that one souldn't be changed in this case ;)

please first upload biosphysio to pypi and then just add it to `"extras"` within `heudiconv/info.py` nearby datalad.

yarikoptic commented on heudiconv/convert.py:
```diff
@@ -407,6 +409,43 @@ def convert_dicom(item_dicoms, bids_options, prefix,
 
+def convert_physio(item_dicoms, bids_options, prefix):
+    """Save DICOM physiology as BIDS physio files
+    Parameters
```
nit: add empty line before Parameters line

yarikoptic commented on heudiconv/dicoms.py:
```diff
@@ -265,8 +265,13 @@ def group_dicoms_into_seqinfos(files, grouping, file_filter=None,
+            # If this is a Siemens PhysioLog, keep it:
+            if mw.dcm_data.SeriesDescription.endswith('_PhysioLog'):
+                # just give it a dummy shape, so that we can continue:
+                mw.image_shape=(0,0,0)
```
nit pep8: spaces around ` = `

pvelasco replied:
Fixed in 9de5114535f08c3dd3d8a177326a920127d1262c

yarikoptic commented on heudiconv/tests/test_regression.py:
```diff
@@ -84,6 +90,32 @@ def test_multiecho(tmpdir, subject='MEEPI', heuristic='bids_ME.py'):
 
 
+@pytest.mark.skipif(not have_bidsphysio, reason="no bidsphysio")
+def test_physio(tmpdir, subject='samplePhysio', heuristic='bids_physio.py'):
```
add 
```
@assert_cwd_unchanged(ok_to_chdir=True)
```
since there is `chdir` used inside

yarikoptic commented on heudiconv/tests/test_regression.py:
```diff
@@ -84,6 +90,32 @@ def test_multiecho(tmpdir, subject='MEEPI', heuristic='bids_ME.py'):
+def test_physio(tmpdir, subject='samplePhysio', heuristic='bids_physio.py'):
+    tmpdir.chdir()
+    outdir = tmpdir.mkdir('out').strpath
+    template = op.join("{subject}/*/*.dcm")
```
what is the point of op.join here if just a single string?

yarikoptic commented on heudiconv/convert.py:
```diff
@@ -407,6 +409,43 @@ def convert_dicom(item_dicoms, bids_options, prefix,
+                "from the last file will be saved."
+            )
+        for dicom_file in item_dicoms:
+            dcm2bidsphysio.dcm2bids( dicom_file, prefix )
```
nit pep8 (?): unneeded spaces at ()?

yarikoptic reviewed (changes requested):
sweet!!!!!  Left notes from a very initial cursory review.

mgxd commented on heudiconv/convert.py:
```diff
@@ -407,6 +409,43 @@ def convert_dicom(item_dicoms, bids_options, prefix,
+    -------
+    None
+    """
+    if bids_options is not None:
+        try:
```
to reduce indentation when possible
```suggestion
    if bids_options is None:
        return
```

pvelasco replied:
Thanks for the suggestion!
Fixed in cbinyu/heudiconv@532c630c4170d4db031e78d8aecde35156eb714b
(I didn't accept your suggested change because I wanted to change the indentation of the function in the same commit).

mgxd commented on heudiconv/convert.py:
```diff
@@ -407,6 +409,43 @@ def convert_dicom(item_dicoms, bids_options, prefix,
+            return
+
+        item_dicoms = list(map(op.abspath, item_dicoms)) # absolute paths
+        if len(item_dicoms)>1:
```
```suggestion
        if len(item_dicoms) > 1:
```

mgxd commented on heudiconv/convert.py:
```diff
@@ -407,6 +409,43 @@ def convert_dicom(item_dicoms, bids_options, prefix,
+        or contain bids specific options
+    prefix : string
+        Conversion outname
+    Returns
```
```suggestion

    Returns
```

mgxd commented on heudiconv/dicoms.py:
```diff
@@ -265,8 +265,13 @@ def group_dicoms_into_seqinfos(files, grouping, file_filter=None,
-            # nothing to see here, just move on
-            continue
+            # If this is a Siemens PhysioLog, keep it:
+            if mw.dcm_data.SeriesDescription.endswith('_PhysioLog'):
```
is this a siemens standard?

pvelasco replied:
Yes.

mgxd commented on heudiconv/dicoms.py:
```diff
@@ -265,8 +265,13 @@ def group_dicoms_into_seqinfos(files, grouping, file_filter=None,
+            # If this is a Siemens PhysioLog, keep it:
+            if mw.dcm_data.SeriesDescription.endswith('_PhysioLog'):
+                # just give it a dummy shape, so that we can continue:
+                mw.image_shape=(0,0,0)
```
```suggestion
                mw.image_shape = (0, 0, 0)
```

mgxd reviewed (commented):
this is awesome, thank you!

a few style suggestions, though we might just consider `black`ing the whole repo and adhering to a stricter style guide.

another consideration while looking at this - WDYT about fetching the data via datalad instead of hosting it in the repo?

yarikoptic commented:
re blacking: I don't mind at all.  But we have outstanding sizeable PRs so it would introduce conflicts... will need to check with @tsalo and @bpinsard

pvelasco commented:
> another consideration while looking at this - WDYT about fetching the data via datalad instead of hosting it in the repo?

Of course, that would be great. You can just download the data from [my branch](https://github.com/cbinyu/heudiconv/tree/dcm_physio/heudiconv/tests/data/samplePhysio), or I can send you a zip file with them.  Once they are in DataLad, I'll modify the regression tests accordingly.

pvelasco committed:
Moves the installation of `bidsphysio` to `info.py`

... from `Dockerfile`

pvelasco committed:
Minor formating in heudiconv/convert.py

Co-authored-by: Mathias Goncalves <goncalves.mathias@gmail.com>

pvelasco committed:
Minor formating in heudiconv/convert.py

Co-authored-by: Mathias Goncalves <goncalves.mathias@gmail.com>

pvelasco committed:
Minor formating in heudiconv/dicoms.py

Co-authored-by: Mathias Goncalves <goncalves.mathias@gmail.com>

tsalo commented:
@yarikoptic, I like automated code formatting, but I think running black would introduce a ton of conflicts in #442. Could aggressive re-formatting be something that's done in its own PR after the pending larger PRs are handled?

mgxd commented:
of course, it would make it nearly unreviewable! I meant showing some love to formatting prior to the next release, and then inserting a style check into our CI

pvelasco committed:
Minor formatting updates

pvelasco committed:
Minor formatting updates

yarikoptic commented:
Just a note (I am ok either way): theoretically (didn't try yet) if there is no value in individual commits in a or (or if they are way to split from a full diff, eg different files), and both branches are auto blacked when pr branch having master branch merged right before blacking - should be easy to force refresh PRs with blacked version ;-) I can try if we decide to black it all

yarikoptic commented:
Re data - I think it is time to remove the middle man me! Let's (ab)use lfs of GitHub with datalad: http://handbook.datalad.org/en/latest/basics/101-138-sharethirdparty.html?highlight=Lfs#use-github-for-sharing-content . I will initiate a repo with the content of this PR data files, and invite everyone possibly interested ;-)...

yarikoptic commented:
Thinking about it - iirc dicoms compress nicely. I will try with enabling shared encryption - that should compress them while stored in lfs.

pvelasco committed:
Adds `assert_cwd_unchanged` to `test_regression`

pvelasco committed:
RF: convert - convert_physio returns if bids_options is None

That reduces indentation in the function code.

pvelasco committed:
ENH: Adds "AcquisitionTime" to the `seqinfo`

pvelasco committed:
Switch back order of seqinfo_fields

To preserve backwards compatibility

pvelasco committed:
Switch back order of SeqInfo arguments in dicom.py

To preserve backwards compatibility

pvelasco committed:
Adds unittest to check "time" in the right position in dicominfo.tsv

pvelasco committed:
ENH: Allows the user to save the Phoenix Report (Siemens) in the sourcedata

This way the user can go back to the scanner and import the exact same protocol that was run for a given subject, improving reproducibility.

pvelasco committed:
Changed calls to bidsphysio to conform with newest version

[`bidsphysio`](https://github.com/cbinyu/bidsphysio) has been refractored into subpackages so that now we only need to import `bidsphysio.dcm2bids`, which has fewer dependencies

pvelasco committed:
Merge branch 'master' into dcm_physio

yarikoptic committed:
BF(py3.5): explicitly case path to str for open

yarikoptic committed:
ENH(minor): sort imports

pvelasco committed:
Merge pull request #6 from cbinyu/adds_acq_time_to_seqinfo

Merges `adds_acq_time_to_seqinfo` into `dcm_physio` branch

pvelasco committed:
Merge branch 'dcm_physio' into handles_phoenix_file

pvelasco committed:
Merge pull request #7 from cbinyu/handles_phoenix_file

Merge `handles_phoenix_file` into `dcm_physio`

pvelasco committed:
Add environment marker (Py>3.5) for dcm2bids requirement.

The dependencies for `bidsphysio.dcm2bids>=1.4.3` require Python 3.6 or higher, so don't try to install `bidsphysio.dcm2bids` unless `python_version > 3.5`.

tsalo commented on heudiconv/info.py:
```diff
@@ -43,7 +43,10 @@
     ],  # Requires patched version ATM ['dcmstack'],
-    'datalad': ['datalad >=%s' % MIN_DATALAD_VERSION]
+    'datalad': ['datalad >=%s' % MIN_DATALAD_VERSION],
+    'dcm2bids': [
```
What about `physio` for this dependency group?

pvelasco replied:
Good idea!

pvelasco committed:
Change name of extra_requires from dcm2bids to physio

Thanks to @tsalo for his suggestion

pvelasco committed:
Delete some unneeded checks in the bids_physio heuristic.

yarikoptic commented on heudiconv/info.py:
```diff
@@ -43,7 +43,10 @@
-    'datalad': ['datalad >=%s' % MIN_DATALAD_VERSION]
+    'datalad': ['datalad >=%s' % MIN_DATALAD_VERSION],
+    'physio': [
+        'bidsphysio.dcm2bids >=1.4.3; python_version>"3.5"',   # if dicoms with physio need to be converted
```
```suggestion
        'bidsphysio.dcm2bids >=1.4.3',   # if dicoms with physio need to be converted
```

as 3.5 is already below what we support

yarikoptic commented on heudiconv/info.py:
```diff
@@ -43,7 +43,10 @@
+    ]
 }
 
 # Flatten the lists
```
all those binary files below adding up to over 1MB -- I think should be moved to some external dataset and used via `fetch_data`. ATM `fetch_data` hardcodes to access from datasets.datalad.org but I think we should change that -- if full url is provided - just use that URL, and then share them on some repo, could be even not annex, just regular git repo on github.

Later we should improve this `fetch_data` to make data persistent locally etc... but not for this PR.

yarikoptic commented on heudiconv/utils.py:
```diff
@@ -46,7 +46,8 @@
     'series_uid',            # 25
- ]
+    'time',                  # 26
+]
```
this would conflict with WiP of #637 / #581 but the main concern is that I do not see it needed/used at all by the target need of this PR -- conversion of physio data. Spurious wishful change? ;-)

yarikoptic reviewed (changes requested):
Sorry it took so long.  Did a pass. Also of interest either bidsphysio can convert fresh WiP 1066 Siemens physio data.

edit: I will convert it to draft. Please undraft/ping whenever ready for re-review etc
